### PR TITLE
src/components/routes: add component RoutesMap for entring routes via a map interface

### PR DESCRIPTION
**Pull request by maceksimon**
Adds map widget based on openlayers library and its Vue 3 integration.

* Install vue3-openlayers (boot file openlayers.js)
* Add RoutesMap - uses library components to enable map features

Add composables:
* useRoutesMap - controls: saving routes, calculating route length, centering map, style function for rendering route paths with vertices.
* useRoutesMapDraw - controls: drawing map features, with the undo function
* useRoutesMapVectorLayer - controls: rendering saved features on the map (without drawing)
* useGeocoding - controls: naming saved routes via the `nominatim.openstreetmap.org` API

### CODE REVIEW

File: src/composables/useRoutesMap.ts [hunk 1]
```diff
@@ -0,0 +1,227 @@
...
+export const useRoutesMap = () => {
+  // provide ol options to child components
+  const isProduction = process.env.NODE_ENV === 'production';
+  const options = {
+    debug: !isProduction,
+  };
+  provide('ol-options', options);
+
+  // constants
+  const MAX_ZOOM_CENTERING_FACTOR = 15;
```
Comment (tmszi) [inline src/composables/useRoutesMap.ts:33]: Use camelCase const name, please.

File: src/composables/useRoutesMap.ts [hunk 2]
```diff
@@ -0,0 +1,227 @@
...
+  const DEFAULT_MAP_ZOOM = 13;
+
+  const mapRef = ref<InstanceType<typeof Map.OlMap> | null>(null);
+  const center = ref<Coordinate>(fromLonLat([14.4378, 50.0755]));
+  const projection = ref('EPSG:3857');
```
Comment (tmszi) [inline src/composables/useRoutesMap.ts:38]: EPSG should be global variable.

File: src/composables/useRoutesMap.ts [hunk 3]
```diff
@@ -0,0 +1,227 @@
...
+    debug: !isProduction,
+  };
+  provide('ol-options', options);
+
+  // constants
+  const MAX_ZOOM_CENTERING_FACTOR = 15;
+  const DEFAULT_MAP_ZOOM = 13;
+
+  const mapRef = ref<InstanceType<typeof Map.OlMap> | null>(null);
+  const center = ref<Coordinate>(fromLonLat([14.4378, 50.0755]));
```
Comment (tmszi) [inline src/composables/useRoutesMap.ts:37]: Map center coordinate should be global variable.

File: src/composables/useRoutesMap.ts [hunk 4]
```diff
@@ -0,0 +1,227 @@
...
+          ]);
+          center.value = coords;
+          if (mapRef.value?.map) {
+            const view = mapRef.value.map.getView();
+            view.setCenter(coords);
+            view.setZoom(DEFAULT_MAP_ZOOM);
+          }
+        },
+        (error) => {
+          console.error('Geolocation error:', error);
```
Comment (tmszi) [inline src/composables/useRoutesMap.ts:151]: User don't know about web dev console. Use modal dialog to show error message, please.

File: src/composables/useRoutesMap.ts [hunk 5]
```diff
@@ -0,0 +1,227 @@
...
+        },
+      );
+    } else {
+      console.error('Geolocation is not supported by this browser.');
```
Comment (tmszi) [inline src/composables/useRoutesMap.ts:155]: Same as my comment above.

File: src/composables/useRoutesMap.ts [hunk 6]
```diff
@@ -0,0 +1,227 @@
...
+          width: 4,
+        }),
+      }),
+    ];
+    if (!geometry) return styles;
+    // style for mid-points
+    styles.push(
+      new Style({
+        geometry: new MultiPoint(geometry.getCoordinates()),
+        image: new Icon({
```
Comment (tmszi) [inline src/composables/useRoutesMap.ts:181]: Do not use emebedded SVG images.

File: src/composables/useRoutesMap.ts [hunk 7]
```diff
@@ -0,0 +1,227 @@
...
+          anchor: [0.5, 0.5],
+        }),
+      }),
+    );
+    // style for starting point
+    styles.push(
+      new Style({
+        geometry: new Point(geometry.getFirstCoordinate()),
+        image: new Icon({
+          src: "data:image/svg+xml,%3Csvg xmlns='http://www.w3.org/2000/svg' width='32' height='32' viewBox='0 0 24 24'%3E%3Cpath fill='currentColor' d='M12 2a7 7 0 0 0-7 7c0 5.25 7 13 7 13s7-7.75 7-13a7 7 0 0 0-7-7m4.5 7H14v5h-4V9H7.5L12 4.5Z'/%3E%3C/svg%3E",
```
Comment (tmszi) [inline src/composables/useRoutesMap.ts:192]: Same as my comment above.

File: src/composables/useRoutesMap.ts [hunk 8]
```diff
@@ -0,0 +1,227 @@
...
+          anchor: [0.5, 1],
+        }),
+      }),
+    );
+    // style for ending point
+    styles.push(
+      new Style({
+        geometry: new Point(geometry.getLastCoordinate()),
+        image: new Icon({
```
Comment (tmszi) [inline src/composables/useRoutesMap.ts:201]: Same as my comment above.

File: src/composables/useRoutesMapTooltip.ts [hunk 9]
```diff
@@ -0,0 +1,73 @@
...
+    // remove drawn route
+    route.value = null;
+    // unset tooltip so that a new one can be created
+    tooltipCoord.value = null;
+    tooltipText.value = '';
+    // cleanup listeners
+    unByKey(listener);
+  };
+
+  const formatGeometryLength = (line: LineString) => {
```
Comment (tmszi) [inline src/composables/useRoutesMapTooltip.ts:61]: Func doc.

File: src/components/__tests__/RoutesMapToolbar.cy.js [hunk 10]
```diff
@@ -0,0 +1,116 @@
...
+    cy.dataCy('toolbar-top').should('be.visible');
+    cy.dataCy('toolbar-bottom').should('be.visible');
+  });
+
+  it('renders draw button', () => {
+    cy.dataCy('add-route-button')
+      .should('be.visible')
+      .and('have.css', 'padding', '0px');
+    cy.dataCy('add-route-avatar').should('be.visible');
+    cy.dataCy('add-route-avatar').invoke('height').should('be.equal', 32);
```
Comment (tmszi) [inline src/components/__tests__/RoutesMapToolbar.cy.js:63]: `const addRouteAvatarIconSize = 32`

Check others repeated icon size number which should be const, please.

File: src/composables/useRoutesMap.ts (hunk 1 above)
Comment (maceksimon) [reply src/composables/useRoutesMap.ts:33]: Updated.

File: src/composables/useRoutesMap.ts (hunk 2 above)
Comment (maceksimon) [reply src/composables/useRoutesMap.ts:38]: Moved to global config.

File: src/composables/useRoutesMap.ts (hunk 3 above)
Comment (maceksimon) [reply src/composables/useRoutesMap.ts:37]: Moved to global config.

File: src/composables/useRoutesMap.ts [hunk 11]
```diff
@@ -0,0 +1,227 @@
...
+  const getRouteLength = (route: Feature | Ref<Feature>): number => {
+    const geom = unref(route).getGeometry();
+    if (geom instanceof LineString) {
+      return getLength(geom);
+    }
+    return 0;
+  };
+
+  /**
+   * Returns a string representing the length of a route in kilometers.
```
[the remark this answers is not in the document]
Comment (maceksimon) [reply src/composables/useRoutesMap.ts:107]: Edited.

File: src/composables/useRoutesMap.ts (hunk 4 above)
Comment (maceksimon) [reply src/composables/useRoutesMap.ts:151]: Replaced console log with Quasar Notify api https://quasar.dev/quasar-plugins/notify#notify-api

File: src/composables/useRoutesMap.ts (hunk 5 above)
Comment (maceksimon) [reply src/composables/useRoutesMap.ts:155]: Replaced console log with Quasar Notify api https://quasar.dev/quasar-plugins/notify#notify-api

File: src/composables/useRoutesMap.ts (hunk 6 above)
Comment (maceksimon) [reply src/composables/useRoutesMap.ts:181]: Replaced with import from /public assets.

File: src/composables/useRoutesMap.ts (hunk 7 above)
Comment (maceksimon) [reply src/composables/useRoutesMap.ts:192]: Replaced with import from /public assets.

File: src/composables/useRoutesMap.ts (hunk 8 above)
Comment (maceksimon) [reply src/composables/useRoutesMap.ts:201]: Replaced with import from /public assets.

File: src/composables/useRoutesMapTooltip.ts (hunk 9 above)
Comment (maceksimon) [reply src/composables/useRoutesMapTooltip.ts:61]: Added.

File: src/components/__tests__/RoutesMapToolbar.cy.js (hunk 10 above)
Comment (maceksimon) [reply src/components/__tests__/RoutesMapToolbar.cy.js:63]: Refactored selectors and variables to constants.

File: src/components/routes/RoutesMap.vue [hunk 12]
```diff
@@ -0,0 +1,390 @@
...
+    // vector layer
+    const vectorLayer = ref<InstanceType<typeof Layers.OlVectorLayer> | null>(
+      null,
+    );
+    const { addMapRoute, clearMapRoutes, renderSavedRoute } =
+      useRoutesMapVectorLayer(vectorLayer);
+
+    /**
+     * Called when a new path is being drawn on the map.
+     * @returns {void}
```
[the remark this answers is not in the document]
Comment (maceksimon) [reply src/components/routes/RoutesMap.vue]: Fixed.

**Comment by tmszi** [comment]
MNK tiles map service URL https://mnk-osm-tiles.dopracenakole.net/{z}/{x}/{y}.png (URL string should be global app var same as OSM URL string)

You can make simple widget (menu with two items) inside map which allow switch between base maps OSM and MNK.

File: src/components/routes/RoutesMap.vue [hunk 13]
```diff
@@ -0,0 +1,385 @@
...
+            :center="center"
+            :zoom="zoom"
+            :projection="projection"
+          />
+          <!-- Layer for OpenStreetMap tiles -->
+          <ol-tile-layer>
+            <ol-source-osm />
+          </ol-tile-layer>
+          <!-- Zoom controls -->
+          <ol-zoom-control zoomInLabel="➕" zoomOutLabel="➖" />
```
Comment (tmszi) [inline src/components/routes/RoutesMap.vue:309]: According [OpenLayers doc ol/control/Zoom](https://openlayers.org/en/latest/apidoc/module-ol_control_Zoom-Zoom.html) `zoomInLabel`  and `zoomOutLabel` attribute has default value.

File: src/components/routes/RoutesMapToolbar.vue [hunk 14]
```diff
@@ -0,0 +1,209 @@
...
+          round
+          unelevated
+          class="q-pa-none q-ma-none"
+          color="transparent"
+          text-color="primary"
+          @click.prevent="$emit('update:draw-enabled', !drawEnabled)"
+          data-cy="add-route-button"
+        >
+          <q-avatar
+            size="32px"
```
Comment (tmszi) [inline src/components/routes/RoutesMapToolbar.vue:75]: `size` attribute value 32px reapet, you can const for this attr

File: src/components/routes/RoutesMapToolbar.vue [hunk 15]
```diff
@@ -0,0 +1,209 @@
...
+            class="q-pa-none q-ma-none"
+            :color="deleteEnabled ? 'primary' : 'grey-3'"
+            data-cy="delete-route-avatar"
+          >
+            <!-- Icon -->
+            <q-icon
+              name="mdi-pencil-remove"
+              :color="deleteEnabled ? 'white' : 'primary'"
+              size="18px"
```
Comment (tmszi) [inline src/components/routes/RoutesMapToolbar.vue:111]: Same as my comment above https://github.com/auto-mat/ride-to-work-by-bike-frontend/pull/470/files#r1682287232

File: src/components/routes/RoutesMapToolbar.vue [hunk 16]
```diff
@@ -0,0 +1,209 @@
...
+      >
+        <q-avatar
+          size="32px"
+          class="q-pa-none q-ma-none"
+          color="white"
+          data-cy="current-position-avatar"
+        >
+          <!-- Icon -->
+          <q-icon
+            name="sym_s_my_location"
```
Comment (tmszi) [inline src/components/routes/RoutesMapToolbar.vue:200]: Unreadable name value

File: src/composables/useGeocoding.ts [hunk 17]
```diff
@@ -0,0 +1,61 @@
...
+   * given its coordinates.
+   *
+   * @param coord number[]
+   * @return Promise<string>
+   */
+  const getLocationName = async (coord: number[]): Promise<string> => {
+    const [lon, lat] = coord;
+    // TODO: update fetch method (if we will use it)
+    const response = await fetch(
+      `https://nominatim.openstreetmap.org/reverse?format=json&lon=${lon}&lat=${lat}`,
```
Comment (tmszi) [inline src/composables/useGeocoding.ts:39]: OSM API URL should be part of global app vars.

File: src/components/routes/RoutesMap.vue [hunk 18]
```diff
@@ -0,0 +1,385 @@
...
+    >
+      <!-- Column: Drawn routes -->
+      <div class="col-12 col-sm-2">
+        <q-scroll-area :style="{ height: listHeight }">
+          <!-- List: Drawn routes -->
+          <q-list separator>
+            <!-- List header -->
+            <q-item class="bg-primary text-white text-weight-bold text-center">
+              <q-item-section class="text-subtitle2 text-uppercase"
+                >Vaše trasy</q-item-section
```
Comment (tmszi) [inline src/components/routes/RoutesMap.vue:252]: Translate string please.

File: src/components/routes/RoutesMap.vue [hunk 19]
```diff
@@ -0,0 +1,385 @@
...
+            <!-- Item: Drawn route -->
+            <q-item
+              clickable
+              v-ripple
+              v-for="(route, index) in savedRoutes"
+              :key="`route-${index}`"
+              @click="onSavedRouteClick(route)"
+            >
+              <q-item-section v-if="route['startName'] && route['endName']">
+                <div>{{ `${route['startName']} → ${route['endName']}` }}</div>
```
Comment (tmszi) [inline src/components/routes/RoutesMap.vue:264]: Better not to use Unicode symbol → here. Use normal icon.

File: src/components/routes/RoutesMap.vue [hunk 20]
```diff
@@ -0,0 +1,385 @@
...
+<template>
+  <div>
+    <!-- Container: Map -->
+    <div
+      class="row q-my-lg"
+      data-cy="routes-map"
+      :style="{
+        borderRadius,
+        overflow: 'hidden',
+        border: '1px solid #E0E0E0',
```
Comment (tmszi) [inline src/components/routes/RoutesMap.vue:241]: We need here extra border color?

File: src/components/routes/RoutesMap.vue (hunk 13 above)
Comment (maceksimon) [reply src/components/routes/RoutesMap.vue:309]: Good point. Using default labels.

File: src/components/routes/RoutesMapToolbar.vue (hunk 14 above)
Comment (maceksimon) [reply src/components/routes/RoutesMapToolbar.vue:75]: Using const.

File: src/components/routes/RoutesMapToolbar.vue (hunk 15 above)
Comment (maceksimon) [reply src/components/routes/RoutesMapToolbar.vue:111]: Updated.

File: src/composables/useGeocoding.ts (hunk 17 above)
Comment (maceksimon) [reply src/composables/useGeocoding.ts:39]: Saved as a global var.

File: src/components/routes/RoutesMap.vue (hunk 18 above)
Comment (maceksimon) [reply src/components/routes/RoutesMap.vue:252]: Translated.

File: src/components/routes/RoutesMap.vue (hunk 20 above)
Comment (maceksimon) [reply src/components/routes/RoutesMap.vue:241]: Replaced with `grey-4`

**Comment by tmszi** [comment]
If you activate one editing tool (draw new line) other tools should be disabled.

**Comment by tmszi** [comment]
1. Simple tooltip above mouse cursor is still missing.
2. Missing switch widget between base maps OSM, MNK

File: src/components/routes/RoutesMapToolbar.vue (hunk 16 above)
Comment (tmszi) [reply src/components/routes/RoutesMapToolbar.vue:237]: ok

**Comment by maceksimon** [comment]
Status
* Added tooltip with date, transport type and route length.
* Added select for layer source.
* Updated undo feature incl. tests (undo now works when tools are switched during drawing and resets after save).

**Comment by tmszi** [comment]
I rebased PR commit https://github.com/auto-mat/ride-to-work-by-bike-frontend/commit/8269a39b75953360aed3d3ad35c03516838f281b  and (installation JS lib packages are|starting dev server) is successful without errors.

However update to new vue3js-openlayers JS lib 11 major version, brings that some map functions stopped working.

**Pull request merged**